solidity/dashboard/src/services/liquidity-rewards.js

```diff
@@ -188,30 +225,128 @@ class SaddleLPRewards extends LiquidityRewards {
   }
 }
 
+class TokenGeyserLPRewards extends LiquidityRewards {
+  static async _getWrappedTokenAddress(LPRewardsContract) {
+    return await LPRewardsContract.methods.token().call()
+  }
+
+  get rewardClaimedEventName() {
+    return "TokensClaimed"
+  }
+
+  get depositWithdrawnEventName() {
+    return "Unstaked"
+  }
+
+  get withdrawTokensFnName() {
+    return "unstake"
+  }
+
+  withdrawTokensArgs(amount) {
+    return [amount, []]
+  }
+
+  stakeArgs(amount) {
+    return [amount, []]
+  }
+
+  stakedBalance = async (address) => {
+    return await this.LPRewardsContract.methods.totalStakedFor(address).call()
+  }
+
+  totalSupply = async () => {
+    return await this.LPRewardsContract.methods.totalStaked().call()
+  }
+
+  rewardBalance = async (address, amount) => {
+    try {
+      // The `TokenGeyser.unstakeQuery` throws an error in case when eg. the
+      // amount param is greater than the real user's stake or when
+      // the user stakes KEEP in block `X` and call unstakeQuery in block `X`
+      // (`SafeMath: division by zero` error is thrown.). The web3 parses the
+      // error message in the wrong way when the `hanleRevert` option is enabled
+      // [1]. So here we clone the rewards contract instance and disable the
+      // `hanldeRevert` option.
+      // References: [1]:
+      // https://github.com/ChainSafe/web3.js/issues/3742
+      const clonedLPRewardsContract = this.LPRewardsContract.clone()
+      clonedLPRewardsContract.handleRevert = false
+      return await clonedLPRewardsContract.methods.unstakeQuery(amount).call()
+    } catch (error) {
+      return 0
```

Comment:
If there's a problem with our call the user will see a 0 reward balance. I wonder if there's a better way to catch this error and render a message to the user. How are we error handling similar cases in the dapp?

Reply:
Good point with error handling. But we don't have any mechanism to display that kind of errors. Also, I'm wondering how can we handle it because we fetch data step by step (I mean we can't get all data in a single request) and sometimes the next value depends on the previous.

So for example, the user is not able to fetch the `deposited KEEP` and `rewarded KEEP` but is able to fetch APY. Should we display a generic error message or specific messages for those values that failed?